cs/test/core/CustomBondedTests.cs

```diff
@@ -0,0 +1,339 @@
+                }
+            }
+        }
+
+        internal class CustomBonded<T, R> : CustomBonded<T>, IBonded<T> 
+            where R : ICloneable<R>
+        {
+            private readonly R reader;
+            private readonly RuntimeSchema schema;
+            private readonly Lazy<T> value;
+
+            public CustomBonded(R reader)
+            {
+                this.reader = reader.Clone();
+                this.schema = RuntimeSchema.Empty;
+
+                this.value = new Lazy<T>(this.Deserialize<T>, LazyThreadSafetyMode.ExecutionAndPublication);
+            }
+
+            public CustomBonded(R reader, RuntimeSchema schema)
+            {
+                this.reader = reader.Clone();
+                this.schema = schema;
+
+                this.value = new Lazy<T>(this.Deserialize<T>, LazyThreadSafetyMode.ExecutionAndPublication);
+            }
+
+            public override T Value
+            {
+                get { return value.Value; }
+            }
+
+            public T Deserialize()
+            {
+                return Deserialize<T>();
+            }
+
+            public void Serialize<W>(W writer)
+            {
+                CustomTransformFactory.Default.Transcoder<R, W>(schema).Transcode(reader.Clone(), writer);
+            }
+
+            public U Deserialize<U>()
+            {
+                return CustomTransformFactory.Default.Deserializer<R, U>(schema).Deserialize<U>(reader.Clone());
+            }
+
+            IBonded<U> IBonded.Convert<U>()
+            {
+                return (IBonded<U>) Convert<U>();
+            }
+
+            public override CustomBonded<U> Convert<U>()
+            {
+                return new CustomBonded<U, R>(reader, schema);
+            }
+        }
+
+        internal class CustomBondedVoid<R> : IBonded
+            where R : ICloneable<R>
+        {
+            private readonly R reader;
+            private readonly RuntimeSchema schema;
+
+            public CustomBondedVoid(R reader, RuntimeSchema schema)
+            {
+                this.reader = reader.Clone();
+                this.schema = schema;
+            }
+
+            public void Serialize<W>(W writer)
+            {
+                CustomTransformFactory.Default.Transcoder<R, W>(schema).Transcode(reader.Clone(), writer);
+            }
+
+            public U Deserialize<U>()
+            {
+                return CustomTransformFactory.Default.Deserializer<R, U>(schema).Deserialize<U>(reader.Clone());
+            }
+
+            IBonded<U> IBonded.Convert<U>()
+            {
+                return new CustomBonded<U, R>(reader, schema);
+            }
+        }
+
+        /// <summary>
+        ///     Custom ITransformFactory for making Deserializer work.
+        /// </summary>
+        private class CustomTransformFactory
+        {
+            public static readonly CustomTransformFactory Default = new CustomTransformFactory();
```

Comment:
- [ ] Minor: `Instance` instead of `Default`, as there's no way to create a custom one of these.